docker/lr-pb/collect_polymerase_read_lengths.py

```diff
@@ -0,0 +1,179 @@
+
+        out_file.write("\t".join(['zmw'] + ['polymerase_read_length']))
+        out_file.write("\n")
+
+        with pysam.AlignmentFile(args.subreads, 'rb', check_sq=False) as bam_file, \
+                pysam.AlignmentFile(args.scraps, 'rb', check_sq=False) as scraps_file:
+
+            num_subreads = 0
+            num_scraps = 0
+
+            subreads_generator = alignment_file_read_generator(bam_file)
+            scraps_generator = alignment_file_read_generator(scraps_file)
+
+            polymerase_read_length_map = dict()
+
+            all_subreads_processed = False
+            all_scraps_processed = False
+
+            while not all_subreads_processed and not all_scraps_processed:
+                # Get info for subreads:
+                if not all_subreads_processed:
+                    all_subreads_processed = update_polymerase_read_count_with_next_read(subreads_generator,
+                                                                                         polymerase_read_length_map)
+                    num_subreads += 1
+
+                # Get info for scraps:
+                if not all_scraps_processed:
+                    all_subreads_processed = update_polymerase_read_count_with_next_read(scraps_generator,
+                                                                                         polymerase_read_length_map)
+                    num_scraps += 1
+
+                if (num_subreads + num_scraps) % 10000 == 0:
+                    LOGGER.info(f"Processed {num_subreads} subreads\t{num_scraps} scraps")
+
+        LOGGER.info(f"Subreads processed: {num_subreads}")
+        LOGGER.info(f"Scraps processed: {num_scraps}")
+        LOGGER.info(f"Total reads processed: {num_subreads + num_scraps}")
+        LOGGER.info(f"Writing results...")
+
+        # Now write out the results:
+        for k,v in polymerase_read_length_map.items():
+            out_file.write(f"{k}\t{v}\n")
+
+    LOGGER.info(f"Results written.")
+
+################################################################################
+
+
+def main(raw_args):
+    base_outfile_name = "polymerase_read_lengths"
+
+    # Get our start time:
+    overall_start = time.time()
+
+    parser = argparse.ArgumentParser(
+        description="Ingests two files: "
+                    "1) PacBio subreads file (SAM/BAM) containing raw subreads off the instrument.  "
+                    "2) PacBio scraps file (SAM/BAM) containing raw scraps off the instrument.  "
+                    "Uses the read names as indicators of polymerase read length (for speed).",
+        usage="Count the lengths of the raw polymerase reads in each zmw.",
+    )
+
+    required_args = parser.add_argument_group("required arguments")
+    required_args.add_argument(
+        "-s", "--subreads", help="PacBio subreads SAM/BAM file.", required=True
+    )
+    required_args.add_argument(
+        "-p", "--scraps", help="PacBio scraps SAM/BAM file.", required=True
```

Comment:
Is there a way to make the scraps file optional?  This script would be useful for the AoU work, but AoU doesn't deliver "scraps.bam" files to us.

Reply:
Sure - I can make it optional, but it will introduce some error in the lengths the script produces.